### ISSUE TITLE
Uncaught exception may not appear on console when main is waiting for a job

### ISSUE THREAD

**Issue by elizarov** [Kotlin/kotlinx.coroutines#208]
Consider the following code:

```
fun main(args: Array<String>) = runBlocking {
    val job = launch {
        doSomething()
    }
    job.join()
}

fun doSomething() {
    TODO("Not implemented")
}
```

Because of the order events are handled internally, `main` can wake up and terminate application before exception handler had a chance to print onto the console from its background thread. 

Consider handling uncaught exceptions in `launch` before everything else.

**Comment by Starcarr** [comment]
This is still happening on `0.22` with Kotlin `1.2.21`.

Code (same as the body of the issue)
```
fun main(args: Array<String>) = runBlocking {
    val job = launch {
        doSomething()
    }
    job.join()
}

fun doSomething() {
    TODO("Not implemented")
}
```
Sample output:
```
/usr/lib/jvm/java-8-openjdk/bin/java...
Process finished with exit code 0
```

**Comment by elizarov** [comment]
Confirm. I've fixed the order in which invocations happen normally, but there is a problem in another place, too.

**Comment by elizarov** [comment]
Here is what I'm observing in this case:
1. The `job` launches, executed and crashes even before the `main` coroutine reaches `job.join` statement.
2. However, the OS scheduler resumes execution on the main thread before the crashed background thread has a change to print its error message on the console. 
3. Note, that by that time the state of the `job` is already complete (we update state first and the print exception on the console), so when main thread reaches `join` it does not even suspend, but goes straight though to the exit, terminates the process and does not leave any chance to the background thread to print its exception.

**Comment by elizarov** [comment]
Fixed in version `0.22.1`. Now, when `launch` Job starts to fail (completes with exception) it starts the process by transitioning into _completing_ state, then it handles the exception (printing its stacktrace by default), then it becomes complete (allowing for a coroutine that is using `join` to go ahead).
